ariadne/src/Ariadne/Wallet/Backend/KeyStorage.hs

```diff
@@ -334,164 +392,186 @@ newWallet walletConfig face runCardanoMode pp mbWalletName mbEntropySize = do
+          , _hdRootHasPassword = hasPass
+          , _hdRootAssurance = assurance
+          , _hdRootCreatedAt = timestamp
+          }
+  -- Map.empty means we're using autogenerated account names
+  throwLeftIO $ update acidDb (CreateHdWallet hdRoot utxoByAccount Map.empty)
+
+  walletRefreshState
+
+-- | convert path in index representation and write it to IORef WalletSelection
+-- TODO:
+-- * check if ui uses the same address indexation as in KeyStorage.hs (Random indexation)
+-- * handle chain type (ui can represent it as new level; proper indexation should be used then)
 select
-  :: WalletFace
+  :: AcidState DB
+  -> WalletFace
   -> IORef (Maybe WalletSelection)
-  -> (CardanoMode ~> IO)
   -> Maybe WalletReference
   -> [Word]
   -> IO ()
-select WalletFace{..} walletSelRef runCardanoMode mWalletRef wsPath = do
-  case mWalletRef of
-    Nothing -> atomicWriteIORef walletSelRef Nothing
+select acidDb WalletFace{..} walletSelRef mWalletRef uiPath = do
+  walletDb <- query acidDb Snapshot
+  let wallets = walletDb ^. dbHdWallets
+  mbSelection <- case mWalletRef of
+    Nothing -> return Nothing
     Just walletRef -> do
-      wsWalletIndex <- resolveWalletRef walletSelRef runCardanoMode walletRef
-      us <- runCardanoMode getSecretDefault
-      -- validate wallet
-      wallet <- maybeThrow
-        (WalletDoesNotExist $ pretty wsWalletIndex)
-        (us ^? usWallets . ix (fromIntegral wsWalletIndex))
-      case nonEmpty wsPath of
-        Nothing -> return ()
+      -- Throw an exception if walletRef is invalid
+      rootId <- resolveWalletRef walletSelRef walletRef walletDb
+      case nonEmpty uiPath of
+        Nothing -> return $ Just $ WSRoot rootId
         Just (accIdx :| acPath) -> do
-
           -- validate account
-          _account <- maybeThrow
+          hdAccount <- maybeThrow
             (AccountDoesNotExist $ pretty accIdx)
-            (wallet ^? wdAccounts . ix (fromIntegral accIdx))
+            ((accList wallets rootId) ^? ix (fromIntegral accIdx))
 
           unless (null acPath) $ throwM SelectIsTooDeep
+          return $ Just $ WSAccount $ hdAccount ^. hdAccountId
 
-      atomicWriteIORef walletSelRef $ Just $
-        WalletSelection { wsPath, wsWalletIndex }
-  walletRefreshUserSecret
+  atomicWriteIORef walletSelRef mbSelection
+  walletRefreshState
+  where
+    -- TODO: make it global
+    accList :: HdWallets -> HdRootId -> [HdAccount]
+    accList wallets rootId = toList $ getAccounts wallets rootId
+
+    getAccounts :: HdWallets -> HdRootId -> IxSet HdAccount
+    getAccounts wallets rootId = fromRight
+      (error "Bug: UnknownHdRoot")
+      (readAccountsByRootId rootId wallets)
 
-getSelectedAddresses :: WalletFace -> IORef (Maybe WalletSelection) -> (CardanoMode ~> IO) -> IO [Address]
-getSelectedAddresses WalletFace{..} walletSelRef runCardanoMode = do
+getBalance
+  :: AcidState DB
+  -> IORef (Maybe WalletSelection)
+  -> IO Coin
+getBalance acidDb walletSelRef = do
```

Comment:
I want a function like `[Address] -> [Coin]`, can you expose it somewhere? So that I can fetch balances for each address of an account AND for the account itself (by summing the list) in one query? It's needed for new Account widget in AD-223.

Reply:
We cache account balances too, no need to sum address balances.

Reply:
The idea is to show a table with all addresses in account and a balance for _each_ of them.

Reply:
AFAIU, we need to add arguments to the `balance` command, so that it's possible to get the balance of something that's not selected (e. g. an address).
Let's do it in a separate issue/PR, because this functionality was present before and this PR merely changes wallet backend implementation without adding new functionality (mostly).